File: kmp/composeApp/src/commonTest/kotlin/com/tanyudii/tmuxweb/domain/FileTreeTest.kt
package com.tanyudii.tmuxweb.domain

import com.tanyudii.tmuxweb.domain.model.ChangedFile
import com.tanyudii.tmuxweb.domain.model.FileStatus
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertTrue

/**
 * Ports ios/TmuxWebClientTests/FileTreeTests.swift 1:1 — same behavior
 * asserted against public/app.js's buildFileTree: folders sort before
 * files, both alphabetically, and nested paths group under a shared folder.
 */
class FileTreeTest {
    private fun file(path: String, status: FileStatus = FileStatus.MODIFIED) =
        ChangedFile(path = path, oldPath = null, status = status, staged = false)

    @Test
    fun `groups nested paths under shared folder`() {
        val tree = buildFileTree(listOf(file("a/b.txt"), file("a/c.txt")))

        assertEquals(1, tree.size)
        assertEquals("a", tree[0].name)
        assertTrue(tree[0].isFolder)
        assertEquals(listOf("b.txt", "c.txt"), tree[0].children.map { it.name })
    }

    @Test
    fun `sorts folders before files alphabetically`() {
        val tree = buildFileTree(listOf(file("z.txt"), file("a/nested.txt"), file("m.txt")))

        assertEquals(listOf("a", "m.txt", "z.txt"), tree.map { it.name })
        assertTrue(tree[0].isFolder)
        assertTrue(!tree[1].isFolder)
    }

    @Test
    fun `root level file has no children and carries its changed file`() {
        val tree = buildFileTree(listOf(file("README.md", status = FileStatus.ADDED)))

        assertEquals(1, tree.size)
        assertTrue(tree[0].children.isEmpty())
        assertEquals(FileStatus.ADDED, tree[0].file?.status)
    }

    @Test
    fun `deeply nested path builds multi level tree`() {
        val tree = buildFileTree(listOf(file("a/b/c/d.txt")))

        assertEquals("a", tree[0].name)
        assertEquals("b", tree[0].children[0].name)
        assertEquals("c", tree[0].children[0].children[0].name)
        assertEquals("d.txt", tree[0].children[0].children[0].children[0].name)
    }

    @Test
    fun `empty path is dropped instead of crashing`() {
        val tree = buildFileTree(listOf(file(""), file("a.txt")))

        assertEquals(1, tree.size)
        assertEquals("a.txt", tree[0].name)
    }
}

File: web/src/domain/fileTree.test.ts
import { describe, expect, test } from "vitest";
import type { ChangedFile, FileStatus } from "../api/types";
import { buildFileTree } from "./fileTree";

// Ports kmp/.../domain/FileTreeTest.kt 1:1 (itself a port of
// ios/TmuxWebClientTests/FileTreeTests.swift / public/app.js's buildFileTree).
function file(path: string, status: FileStatus = "modified"): ChangedFile {
  return { path, status, staged: false, conflicted: false };
}

describe("buildFileTree", () => {
  test("groups nested paths under shared folder", () => {
    const tree = buildFileTree([file("a/b.txt"), file("a/c.txt")]);

    expect(tree).toHaveLength(1);
    expect(tree[0].name).toBe("a");
    expect(tree[0].isFolder).toBe(true);
    expect(tree[0].children.map((child) => child.name)).toEqual(["b.txt", "c.txt"]);
  });

  test("sorts folders before files alphabetically", () => {
    const tree = buildFileTree([file("z.txt"), file("a/nested.txt"), file("m.txt")]);

    expect(tree.map((node) => node.name)).toEqual(["a", "m.txt", "z.txt"]);
    expect(tree[0].isFolder).toBe(true);
    expect(tree[1].isFolder).toBe(false);
  });

  test("root level file has no children and carries its changed file", () => {
    const tree = buildFileTree([file("README.md", "added")]);

    expect(tree).toHaveLength(1);
    expect(tree[0].children).toEqual([]);
    expect(tree[0].file?.status).toBe("added");
  });

  test("deeply nested path builds multi level tree", () => {
    const tree = buildFileTree([file("a/b/c/d.txt")]);

    expect(tree[0].name).toBe("a");
    expect(tree[0].children[0].name).toBe("b");
    expect(tree[0].children[0].children[0].name).toBe("c");
    expect(tree[0].children[0].children[0].children[0].name).toBe("d.txt");
  });

  test("empty path is dropped instead of crashing", () => {
    const tree = buildFileTree([file(""), file("a.txt")]);

    expect(tree).toHaveLength(1);
    expect(tree[0].name).toBe("a.txt");
  });
});
